src/DatumCacheWsp.purs

```diff
@@ -0,0 +1,220 @@
+module DatumCacheWsp
+  ( DatumCacheResponse(..)
+  , DatumCacheRequest(..)
+  , DatumCacheMethod(..)
+  , JsonWspRequest
+  , JsonWspResponse
+  , WspFault(WspFault)
+  , faultToString
+  , jsonWspRequest
+  , parseJsonWspResponse
+  , responseMethod
+  , requestMethodName
+  ) where
+
+import Control.Alt (map, (<$), (<$>), (<|>))
+import Control.Bind ((=<<), bind)
+import Control.Category ((<<<))
+import Data.Argonaut
+  ( Json
+  , JsonDecodeError(UnexpectedValue)
+  , caseJsonObject
+  , decodeJson
+  , encodeJson
+  , jsonNull
+  , stringify
+  )
+import Data.Array (foldM)
+import Data.Bifunctor (lmap)
+import Data.Either (Either(Left, Right), note)
+import Data.Eq (class Eq)
+import Data.Function ((>>>), const, ($))
+import Data.Maybe (Maybe(Just, Nothing), maybe)
+import Data.Newtype (unwrap)
+import Data.Traversable (traverse)
+import Data.Unit (Unit, unit)
+import Foreign.Object as FO
+import Serialization.Address (BlockId, Slot)
+import Types.ByteArray (byteArrayToHex, hexToByteArray)
+import Types.PlutusData (PlutusData, DatumHash)
+import Types.Transaction (DataHash(DataHash))
+
+newtype WspFault = WspFault Json
+
+faultToString :: WspFault -> String
+faultToString (WspFault j) = stringify j
+
+type JsonWspRequest =
+  { type :: String
+  , version :: String
+  , servicename :: String
+  , methodname :: String
+  , args :: Json
+  }
+
+type JsonWspResponse =
+  { type :: String
+  , version :: String
+  , servicename :: String
+  , methodname :: String
+  , result :: Maybe Json
+  , fault :: Maybe Json
+  }
+
+data DatumCacheMethod
+  = GetDatumByHash
+  | GetDatumsByHashes
+  | StartFetchBlocks
+  | CancelFetchBlocks
+  | DatumFilterAddHashes
+  | DatumFilterRemoveHashes
+  | DatumFilterSetHashes
+  | DatumFilterGetHashes
+
+derive instance Eq DatumCacheMethod
+
+datumCacheMethodToString :: DatumCacheMethod -> String
```

Comment:
Why not use a `Show` instance with `genericShow` for this?

Reply:
I'd have to restructure fromString to match on genericShow result. I can do it, but I find this approach just more straightforward.